lib/Dialect/TTIR/IR/TTIROps.cpp

```diff
@@ -1617,6 +1715,56 @@ ::mlir::LogicalResult mlir::tt::ttir::PermuteOp::verify() {
   return success();
 }
 
+// PermuteOp canonicalization
+::mlir::LogicalResult
+mlir::tt::ttir::PermuteOp::canonicalize(PermuteOp op,
+                                        PatternRewriter &rewriter) {
+  // If the permutation is the identity permutation, then the operation is a
+  // no-op and can be replaced with the input tensor.
+  if (llvm::is_sorted(op.getPermutation())) {
+    rewriter.replaceOp(op, op.getInput());
+    return success();
+  }
+
+  // If there are two consecutive permutations check if they can be merged
+  // into single permutation or removed.
+  if (auto inputPermuteOp = op.getInput().getDefiningOp<PermuteOp>()) {
+    size_t numUsers = std::distance(inputPermuteOp.getOperation()->use_begin(),
+
+                                    inputPermuteOp.getOperation()->use_end());
+    // If number of users of the input permutation is more than 1, then we
+    // cannot merge or remove the permutation.
+    if (numUsers > 1) {
```

Comment:
Yea you are correct. Regarding erase I like to erase even if there is DCE pass since it makes reading intermediate IR a lot easier for me. But I will skip erasing here since it makes most sense.